app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js

```diff
@@ -145,6 +155,76 @@ async function addEthereumChainHandler(
     );
   }
 
+  const switchChainWithPermissions = async ({
+    networkConfigurationId,
+    approvalFlowId,
+  }) => {
+    let permissionedChainIds;
+    try {
+      ({ value: permissionedChainIds } = getCaveat(
+        origin,
+        PermissionNames.wallet_switchEthereumChain,
+        CaveatTypes.restrictNetworkSwitching,
+      ));
+    } catch (e) {
+      // throws if the origin does not have any switchEthereumChain permissions yet
+      if (e instanceof PermissionDoesNotExistError) {
+        // suppress expected error in case that the domain does not have a
+        // wallet_switchEthereumChain permission set yet
+      } else {
+        throw e;
+      }
+    }
```

Comment:
Consider migrating this logic to the `getCaveat` hook itself.

Reply:
done here: https://github.com/MetaMask/metamask-extension/pull/24415/commits/eb4393e50321836c4cef87206b0abd3895e26993